src/lib/network_pool/network_pool_base.ml

```diff
@@ -97,21 +102,36 @@ end)
         Broadcast_callback.error e cb
 
   let process_incoming_resource_pool_diff t diff cb =
-    match%bind Resource_pool.Diff.verify t.resource_pool diff with
-    | Error err ->
-        [%log' info t.logger] "Refusing to rebroadcast. Verification error: %s"
-          (Error.to_string_hum err) ;
-        Broadcast_callback.error err cb
-    | Ok verified_diff ->
-        [%log' debug t.logger] "Verified diff: $verified_diff"
-          ~metadata:
-            [ ( "verified_diff"
-              , Resource_pool.Diff.verified_to_yojson
-                @@ Envelope.Incoming.data verified_diff )
-            ; ( "sender"
-              , Envelope.Sender.to_yojson
-                @@ Envelope.Incoming.sender verified_diff ) ] ;
-        apply_and_broadcast t verified_diff cb
+    [%log' debug t.logger] "Verifying $diff"
+      ~metadata:
+        [("diff", Resource_pool.Diff.to_yojson @@ Envelope.Incoming.data diff)] ;
+    let r, w =
```

Comment:
Why the pipe here? Won't there only ever be one thing in this pipe?

Reply:
so that we don't wait for `apply_and_rebroadcast` when iterating the merge pipe below? maybe that's not required

Reply:
This will make a new pipe for each invocation of this, but this isn't what we want. We want 1 pipe behind the merge pipe, not 1 pipe behind the merge pipe per diff we want to process.

To fix this, move the pipe creation logic into `of_resource_pool_and_diffs` and setup the merge pipe to write to it instead of calling `process_incoming_resource_pool_diff`. Then the in-betwixt pipe will call `don't_wait_for (let%bind res = process_incoming_resource_pool_diff t diff cb in match res with | Ok (diff, cb) -> apply_and_broadcast t diff cb | Error _ -> ())` or something like that.

We should still even test that well to make sure we aren't introducing a race condition into the network pool. If we see issues, we will need to rethink how we implemented snark verification batching some, and find a better way to fit that batching into the network pool abstraction.

Reply:
As per the discussion with @nholland94 changing this to filter the diffs asynchronously and applying the verified diffs synchronously